packages/core/contracts/oracle/implementation/SkinnyOptimisticOracle.sol

```diff
@@ -0,0 +1,658 @@
+     * @param _customLiveness custom proposal liveness to set for request.
+     * @return totalBond default bond + final fee that the proposer and disputer will be required to pay.
+     */
+    function requestPrice(
+        bytes32 _identifier,
+        uint256 _timestamp,
+        bytes memory _ancillaryData,
+        IERC20 _currency,
+        uint256 _reward,
+        uint256 _bond,
+        uint256 _customLiveness
+    ) external override nonReentrant() returns (uint256 totalBond) {
+        bytes32 requestId = _getId(msg.sender, _identifier, _timestamp, _ancillaryData);
+        require(requests[requestId] == bytes32(0), "Request already initialized");
+        require(_getIdentifierWhitelist().isIdentifierSupported(_identifier), "Unsupported identifier");
+        require(_getCollateralWhitelist().isOnWhitelist(address(_currency)), "Unsupported currency");
+        require(_timestamp <= getCurrentTime(), "Timestamp in future");
+        require(
+            _stampAncillaryData(_ancillaryData, msg.sender).length <= ancillaryBytesLimit,
+            "Ancillary Data too long"
+        );
+        uint256 finalFee = _getStore().computeFinalFee(address(_currency)).rawValue;
+
+        // Associate new request with ID
+        Request memory request;
+        request.currency = _currency;
+        request.reward = _reward;
+        request.finalFee = finalFee;
+        request.bond = _bond != 0 ? _bond : finalFee;
+        request.customLiveness = _customLiveness;
+
+        if (_reward > 0) {
+            _currency.safeTransferFrom(msg.sender, address(this), _reward);
+        }
+
+        _storeRequestHash(requestId, request);
+        emit RequestPrice(msg.sender, _identifier, _timestamp, _ancillaryData, request);
+
+        return request.bond.add(request.finalFee);
+    }
+
+    /**
+     * @notice Proposes a price value on another address' behalf. Note: this address will receive any rewards that come
+     * from this proposal. However, any bonds are pulled from the caller.
+     * @param _requester sender of the initial price request.
+     * @param _identifier price identifier to identify the existing request.
+     * @param _timestamp timestamp to identify the existing request.
+     * @param _ancillaryData ancillary data of the price being requested.
+     * @param _request price request parameters whose hash must match the request that the caller wants to
+     * propose a price for.
+     * @param _proposer address to set as the proposer.
+     * @param _proposedPrice price being proposed.
+     * @return totalBond the amount that's pulled from the caller's wallet as a bond. The bond will be returned to
+     * the proposer once settled if the proposal is correct.
+     */
+    function proposePriceFor(
+        address _requester,
+        bytes32 _identifier,
+        uint256 _timestamp,
+        bytes memory _ancillaryData,
+        Request memory _request,
```

Comment:
Not a gas optimization, but was curious if you think it would be simpler to bundle these params into a struct:
```
        address _requester,
        bytes32 _identifier,
        uint256 _timestamp,
        bytes memory _ancillaryData,
        Request memory _request,
```

I only say this because they're passed around a lot and emitted together in a bunch of events. Definitely not strongly opinionated though.

Reply:
I think keeping them separate as params makes the interface a bit easier to work with externally personally.

Reply:
SGTM